### PR TITLE
[AMORO-1457]Support computed columns  and watermark in Flink DDL 

### LINKED ISSUE #1457: [Improvement]: Support computed columns in Flink DDL
### Search before asking

- [X] I have searched in the [issues](https://github.com/NetEase/arctic/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

```sql
create table arctic.test.L (id int, proc as PROCTIME())
```

```java
Caused by: java.lang.UnsupportedOperationException: Creating table with computed columns is not supported yet.
	at com.netease.arctic.flink.catalog.ArcticCatalog.lambda$validateFlinkTable$3(ArcticCatalog.java:328)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at com.netease.arctic.flink.catalog.ArcticCatalog.validateFlinkTable(ArcticCatalog.java:326)
	at com.netease.arctic.flink.catalog.ArcticCatalog.createTable(ArcticCatalog.java:271)
```

### How should we improve?

Filter the computed column or watermark definition field.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Subtasks

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by xieyi888**
## Why are the changes needed?

Close #1457 .

## Brief change log

1. suppoort serialized  compute columns and watermarks into arctic table properties
compute columns and watermark in arctic properties like below
image

2. support deserializing compute columns and watermarks into flink tableSchema
3. Only convert flink physicalColums into arctic(iceberg) schema during create table, compute columns and watermarks should be serialized into arctic table properties
4. when get table(write and read), deserializing compute columns and watermarks into flink TableSchema ,and combine with flink TableSchema convert by arctic(iceberg) schema
5. only support one watermark strategy
6. not support alter properties for compute columns and watermark strategy now.

## How was this patch tested?

- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

Local test includes as below:
**flink 1.15, mix_iceberg**
1.1 create arctic table with compute columns and watermark under flink arctic catalog.  

1.2 read and write table with compute columns.
```sql

        tableEnv.executeSql("CREATE CATALOG arctic_test WITH ( " +
                "  'type'='arctic', " +
                "  'metastore.url' = 'thrift://****:1260/mix_iceberg'  " + 
                ")");

       tableEnv.executeSql("CREATE TABLE arctic_test.iceberg_test_db.arctic_catalog_mix_iceberg_column2_orc (  " +
               "                     `id` int,    " +
               "                      `cal_id` as id+5, "  +
               "                       proc as PROCTIME() " +
               ") ");

       tableEnv.executeSql("CREATE TABLE gen_order (  " +
               "id INT " +
               ") " +
               "with( " +
               "'connector' = 'datagen',  " +
               "'rows-per-second'='1',  " +
               "'fields.id.kind'='random',  " +
               "'fields.id.min'='1',  " +
               "'fields.id.max'='1000') ");

       tableEnv.executeSql("INSERT INTO  arctic_test.iceberg_test_db.arctic_catalog_mix_iceberg_column2_orc       " +
               "                      select id          " +
               "                   from gen_order where id<10");

        tableEnv.sqlQuery("select * from arctic_test.iceberg_test_db.arctic_catalog_mix_iceberg_column2_orc")
                .execute().print();

```
1.3 read  arctic table and do lookup join with mysql dimension table
```sql

        tableEnv.executeSql("CREATE CATALOG arctic_test WITH ( " +
                "  'type'='arctic', " +
                "  'metastore.url' = 'thrift://****:1260/mix_iceberg'  " + 
                ")");

        tableEnv.executeSql("CREATE TABLE dim_table(    " +
                "id bigint,    " +
                "name string,    " +
                "PRIMARY KEY(id) NOT ENFORCED    " +
                ") WITH (    " +
                "  'connector' = 'jdbc',    " +
                "  'url' = 'jdbc:mysql://****:3306/sloth_learning',    " +
                "  'table-name' = 'dim_table',    " +
                "  'username' = '****',    " +
                "  'password' = '****'    " +
                ")");

        tableEnv.sqlQuery("select t1.id, t1.cal_id,t1.proc,t2.name from     " +
                "arctic_test.iceberg_test_db.arctic_catalog_mix_iceberg_column2_orc t1    " +
                "JOIN dim_table FOR SYSTEM_TIME AS OF t1.proc as t2 on t1.id=t2.id")
                .execute().print();
```

2.1. create arctic table with compute columns under flink default catalog.  read  with compute columns.
```sql
        tableEnv.executeSql(" CREATE TABLE flink_catalog_mix_iceberg_column1_orc (         " +
                "                                      id bigint,           " +
                "                                      `cal_id` as id+5,     " +
                "                                      proc as PROCTIME()    " +
                "                                      " +
                "                 ) WITH (         " +
                "                     'connector' = 'arctic',         " +
                "                     'metastore.url' = 'thrift://****:1260/mix_iceberg',         " +
                "                     'arctic.catalog' = 'mix_iceberg',         " +
                "                     'arctic.database' = 'iceberg_test_db',         " +
                "                     'arctic.table' = 'flink_catalog_mix_iceberg_column1_orc'         " +
                "                 )");


        tableEnv.sqlQuery("select * from flink_catalog_mix_iceberg_column1_orc")
                .execute().print();
```
2.2 read  arctic table and do lookup join with mysql dimension table
```sql

        tableEnv.executeSql(" CREATE TABLE flink_catalog_mix_iceberg_column1_orc (        " +
                "                                      id bigint,          " +
                "                                      `cal_id` as id+5,    " +
                "                                      proc as PROCTIME()   " +
                "                                     " +
                "                 ) WITH (        " +
                "                     'connector' = 'arctic',        " +
                "                     'metastore.url' = 'thrift://****:1260/mix_iceberg',        " +
                "                     'arctic.catalog' = 'mix_iceberg',        " +
                "                     'arctic.database' = 'iceberg_test_db',        " +
                "                     'arctic.table' = 'flink_catalog_mix_iceberg_column1_orc'        " +
                "                 )");

        tableEnv.executeSql("CREATE TABLE dim_table(   " +
                "id bigint,   " +
                "name string,   " +
                "PRIMARY KEY(id) NOT ENFORCED   " +
                ") WITH (   " +
                "  'connector' = 'jdbc',   " +
                "  'url' = 'jdbc:mysql://****:3306/****',   " +
                "  'table-name' = 'dim_table',   " +
                "  'username' = '****',   " +
                "  'password' = '****'   " +
                ")");

        tableEnv.sqlQuery("select t1.id, t1.cal_id,t1.proc,t2.name from    " +
                "flink_catalog_mix_iceberg_column1_orc t1   " +
                "JOIN dim_table FOR SYSTEM_TIME AS OF t1.proc as t2 on t1.id=t2.id")
                .execute().print();
``` 

## Documentation

- Does this pull request introduce a new feature? (yes / no)
yes
- If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
docs

### CODE REVIEW

**Comment by xieyi888** [comment]
some  unit test failed, I'll try to fix it

**Comment by YesOrNo828** [comment]
@xieyi888 there are some UTs failed about the amoro's schema fetching, could you kindly have a time to take a look?
com.netease.arctic.flink.table.TestJoin.testLookupJoinWithPartialFields
TestJoin.testLookupJoin:221->FlinkTestBase.exec:203->FlinkTestBase.exec:207 » IllegalArgument Field op_time not found in source schema
TestJoin.testLookupJoinWithPartialFields:316->FlinkTestBase.exec:203->FlinkTestBase.exec:207 » IndexOutOfBounds Index: 0, Size: 0
TestJoin.testRightEmptyLookupJoin:130->FlinkTestBase.exec:203->FlinkTestBase.exec:207 » IllegalArgument Field op_time not found in source schema

**Comment by xieyi888** [comment]
> @xieyi888 there are some UTs failed about the amoro's schema fetching, could you kindly have a time to take a look?
> 
> TestJoin.testLookupJoin:221->FlinkTestBase.exec:203->FlinkTestBase.exec:207 » IllegalArgument Field op_time not found in source schema TestJoin.testLookupJoinWithPartialFields:316->FlinkTestBase.exec:203->FlinkTestBase.exec:207 » IndexOutOfBounds Index: 0, Size: 0 TestJoin.testRightEmptyLookupJoin:130->FlinkTestBase.exec:203->FlinkTestBase.exec:207 » IllegalArgument Field op_time not found in source schema

Thanks , I had  separate function getPhysicalSchemaForDimTable when dim_table.enabled=true. TestJoin can work normally

**Comment by xieyi888** [comment]
@YesOrNo828 @zhoujinsong zhou Could you please review this pr

File: flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java [hunk 1]
```diff
@@ -18,49 +18,100 @@
...
    *
-   * @param rowType a RowType
-   * @param primaryKeys
+   * @param arcticSchema
+   * @param arcticProperties
    * @return Flink TableSchema
    */
-  public static TableSchema toSchema(RowType rowType, List<String> primaryKeys) {
+  public static TableSchema toSchema(
+      Schema arcticSchema, List<String> primaryKeys, Map<String, String> arcticProperties) {
```
Comment (YesOrNo828) [inline flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java:78]:
```suggestion
      Schema amoroSchema, List<String> primaryKeys, Map<String, String> amoroProperties) {
```
PLS also renames the word `arctic` to `amoro` in the comments.

File: flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java [hunk 2]
```diff
@@ -197,4 +245,122 @@ public static void addPrimaryKey(
                               "Arctic primary key should be declared in table")));
         });
   }
+  /**
+   * Generate table properties for watermark and computed columns from flink TableSchema.
+   *
+   * @param schema flink TableSchema.
```
Comment (YesOrNo828) [inline flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java:251]:
```suggestion
   * @param schema Flink TableSchema.
```

File: flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java [hunk 3]
```diff
@@ -197,4 +245,122 @@ public static void addPrimaryKey(
...
+   * @return Table properties map.
+   */
+  public static Map<String, String> addSchemaProperties(TableSchema schema) {
```
Comment (YesOrNo828) [inline flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java:254]: How about 'generateExtraOptionsFrom'?
```suggestion
  public static Map<String, String> generateExtraOptionsFrom(TableSchema schema) {
```

File: flink/flink-common/src/main/java/com/netease/arctic/flink/catalog/ArcticCatalog.java [hunk 4]
```diff
@@ -328,7 +328,13 @@ public void createTable(ObjectPath tablePath, CatalogBaseTable table, boolean ig
     PartitionSpec spec = toPartitionSpec(((CatalogTable) table).getPartitionKeys(), icebergSchema);
     tableBuilder.withPartitionSpec(spec);
 
-    tableBuilder.withProperties(table.getOptions());
+    Map<String, String> properties = table.getOptions();
+    // update computed columns and watermark to properties
+
+    Map<String, String> updatedProperties = addSchemaProperties(tableSchema);
```
Comment (YesOrNo828) [inline flink/flink-common/src/main/java/com/netease/arctic/flink/catalog/ArcticCatalog.java:334]:
```suggestion
    Map<String, String> extraOptions = addSchemaProperties(tableSchema);
```

File: flink/flink-common/src/main/java/com/netease/arctic/flink/table/DynamicTableFactory.java [hunk 5]
```diff
@@ -145,8 +147,13 @@ public DynamicTableSource createDynamicTableSource(Context context) {
       objectPath = new ObjectPath(identifier.getDatabaseName(), identifier.getObjectName());
     }
 
+    // update computed columns and watermark to properties
+    Map<String, String> properties = options.toMap();
+    Map<String, String> updatedProperties = addSchemaProperties(catalogTable.getSchema());
```
Comment (YesOrNo828) [inline flink/flink-common/src/main/java/com/netease/arctic/flink/table/DynamicTableFactory.java:152]: ditto

File: flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java [hunk 6]
```diff
@@ -197,4 +245,122 @@ public static void addPrimaryKey(
...
+                  ((TableColumn.ComputedColumn) column).getExpression());
+            });
+
+    return serialized;
+  }
+
+  /** deserialize computeColumns from properties */
+  private static TableColumn deserializeComputeColumn(
+      Map<String, String> arcticProperties, int index) {
+    String name = arcticProperties.get(compoundKey(FLINK_PREFIX, COMPUTED_COLUMNS, index, NAME));
```
Comment (YesOrNo828) [inline flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java:306]: I think we should check if the column name exists in the Amoro Schema. 
If the column name has been renamed, but the options didn't update. We should throw an exception when meeting this case.
And we should add a corner test for it.

**Review by YesOrNo828 — COMMENTED** [review]
@xieyi888  Thanks for the PR, I left some comments, please kindly have a look.

File: flink/flink-common/src/main/java/com/netease/arctic/flink/catalog/ArcticCatalog.java [hunk 7]
```diff
@@ -289,29 +287,31 @@ public void renameTable(ObjectPath tablePath, String newTableName, boolean ignor
...
+    createArcticTable(tablePath, table, ignoreIfExists);
+  }
 
+  private void createArcticTable(
+      ObjectPath tablePath, CatalogBaseTable table, boolean ignoreIfExists)
+      throws CatalogException, TableAlreadyExistException {
     TableSchema tableSchema = table.getSchema();
     TableSchema.Builder flinkSchemaBuilder = TableSchema.builder();
-
+    // add PhysicalColumn to arcticTableSchema
```
Comment (zhoujinsong) [inline flink/flink-common/src/main/java/com/netease/arctic/flink/catalog/ArcticCatalog.java:298]: Maybe we can generate the physical schema with `com.netease.arctic.flink.FlinkSchemaUtil#getPhysicalSchema`.

**Review by zhoujinsong — COMMENTED** [review]
@xieyi888 Thanks a lot for your work!
I left some comments. Please take another look.

**Comment by zhoujinsong** [comment]
BTW, I'm curious about the property design of compute columns. The index in the property key doesn't seem to have a specific use. Would a design like this be better for users to understand these properties, such as:
```
flink.compute-column.cal_id.data_type=INT, 
flink.compute-column.cal_id.expr=`id`+5
```

File: flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java (hunk 6 above)
Comment (xieyi888) [reply flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java:306]: > I think we should check if the column name exists in the Amoro Schema. If the column name has been renamed, but the table options weren't updated. We should throw an exception when meeting this case. And we should add a corner test for it.

@YesOrNo828  I think it's necessary.
 Do you mean the phenomenon below?
 1. Flink DDL  
physical colum id and compute column cal_id
 ```
CREATE TABLE mix_iceberg.test_db.test_db (  
        `id` int,          
        `cal_id` as id+5,        
        proc as PROCTIME()       
        )  
```
tableProperties as below
```
flink.computed-column.1.expr= `id` + 5
flink.computed-column.1.name= cal_id
flink.computed-column.1.data-type= INT
```

2. other engine, such spark rename `id` into `newid`
`ALTER TABLE mix_iceberg.test_db.test_db RENAME COLUMN id TO newid;`

3. deserializeComputeColumn function should throw exception if tableProperties weren't updated for `flink.computed-column.1.xx`
4. deserializeComputeColumn function work normaly, after column renamed and tableProperties are updated

**Comment by xieyi888** [comment]
> BTW, I'm curious about the property design of compute columns. The index in the property key doesn't seem to have a specific use. Would a design like this be better for users to understand these properties, such as:
> 
> ```
> flink.compute-column.cal_id.data_type=INT, 
> flink.compute-column.cal_id.expr=`id`+5
> ```
> 
> HDYT? @xieyi888 @YesOrNo828

The index is depedended on flink TableSchema index.  It just to specify the different columns. Although It seems not very convenient for users to understand the properties. It's better for serialize/deserialize When the field name contains some exception string. HDYT? @YesOrNo828

**Comment by zhoujinsong** [comment]
> It's better for serialize/deserialize When the field name contains some exception string. 

Yes, Including column names in property names can indeed introduce some special characters.

So perhaps we can still use an index to differentiate different compute columns. The index would only count the number of compute columns, starting from 1, and we would maintain this order when reassembling the schema. This might make it easier for users to understand.

**Comment by YesOrNo828** [comment]
I agree with using the index to determine the order of fields(including physical and computed columns) instead of the field name.

There would be a data type mismatching exception. If we only use an index to differentiate different compute columns, not include physical columns.
For example:
```
CREATE TABLE amoro.db.T ( 
  id int,
  id2 as id + 2,
  id3 as id * 2,
  name string);

INSERT INTO R
SELECT * FROM T;
```
The query of `select * from T` would return the schema: id, name, id2, id3.  But this is a disorder with the original table R's data types(int, int, int, name).

File: flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java [hunk 8]
```diff
@@ -18,49 +18,103 @@
...
+   * Convert iceberg Schema to flink TableSchema.
    *
-   * @param rowType a RowType
-   * @param primaryKeys
+   * @param icebergSchema
+   * @param tableProperties
    * @return Flink TableSchema
    */
-  public static TableSchema toSchema(RowType rowType, List<String> primaryKeys) {
+  public static TableSchema toSchema(
```
Comment (YesOrNo828) [inline flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java:84]: So according to the prefix comment, I would suggest building the Flink table schema by iterating the column count.
e.g.:
```
for(int i : columnCount) {
    // build index=i column
    // The index=i column may be a physical column, watermark, computed column, or metadata column. 
}

**Comment by xieyi888** [comment]
> I agree with using the index to determine the order of fields(including physical and computed columns) instead of the field name.
> 
> There would be a data type mismatching exception. If we only use an index to differentiate different compute columns, not include physical columns. For example:
> 
> ```
> CREATE TABLE amoro.db.T ( 
>   id int,
>   id2 as id + 2,
>   id3 as id * 2,
>   name string);
> 
> INSERT INTO R
> SELECT * FROM T;
> ```
> 
> The query of `select * from T` would return the schema: id, name, id2, id3. But this is a disorder with the original table R's data types(int, int, int, name).

I agree with it.  In order to make the same for create and read table with columns(physical column,  computed column, metadata column. ) and watermark
so Maybe we can design as below, and make somn documents for users.

1. when create table with DDL,  We serialized  columns(expect physical column) into table properties as below, ${index} means the column index(order) in table.
```
flink.schema.column.${index}.name= **
flink.schema.column.${index}.data-type=**
```
if it's computed column, we also serialized  the expresion
```
flink.schema.column.${index}.expr=**
```
if it's  computed column, we also serialized  metadata and virtual
```
flink.schema.column.${index}.metadata=**
flink.schema.column.${index}.virtual=**
```
2. when create table with DDL,  We serialized watermark into table properties as below
```
flink.schema.watermark.rowtime=**
flink.schema.watermark.strategy.expr=**
flink.schema.watermark.strategy.data-type=**
```

HDYT @YesOrNo828 @zhoujinsong

File: flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java (hunk 8 above)
Comment (xieyi888) [reply flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java:84]: > So according to the prefix comment, I would suggest building the Flink table schema by iterating the column count. e.g.:
> 
> ```
> for(int i : columnCount) {
>     // build index=i column
>     // The index=i column may be a physical column, watermark, computed column, or metadata column. 
> }
> ```

OK, I'll change it

**Comment by xieyi888** [comment]
> > It's better for serialize/deserialize When the field name contains some exception string.
> 
> Yes, Including column names in property names can indeed introduce some special characters.
> 
> So perhaps we can still use an index to differentiate different compute columns. The index would only count the number of compute columns, starting from 1, and we would maintain this order when reassembling the schema. This might make it easier for users to understand.

 if physical column was deleted by spark or other engine, property for compute column would be destroed and leading to deserializing compute column inaccuracy, we deside to make compute column  listed after all physical columns and validate column order during create table

File: flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java (hunk 8 above)
Comment (xieyi888) [reply flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java:84]: > So according to the prefix comment, I would suggest building the Flink table schema by iterating the column count. e.g.:
> 
> ```
> for(int i : columnCount) {
>     // build index=i column
>     // The index=i column may be a physical column, watermark, computed column, or metadata column. 
> }
> ```


As we will make compute column at last, and start from 1, will keep use set for compute indexs

**Comment by xieyi888** [comment]
> > > It's better for serialize/deserialize When the field name contains some exception string.
> > 
> > 
> > Yes, Including column names in property names can indeed introduce some special characters.
> > So perhaps we can still use an index to differentiate different compute columns. The index would only count the number of compute columns, starting from 1, and we would maintain this order when reassembling the schema. This might make it easier for users to understand.
> 
> if physical column was deleted by spark or other engine, property for compute column would be destroed and leading to deserializing compute column inaccuracy, we deside to make compute column listed after all physical columns and validate column order during create table

done

File: flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java [hunk 9]
```diff
@@ -197,4 +242,191 @@ public static void addPrimaryKey(
...
+      if (watermarkSpecs.size() > 1) {
+        throw new IllegalStateException("Multiple watermark definition is not supported yet.");
+      }
+      properties.putAll(serializeWatermarkSpec(watermarkSpecs.get(0)));
+    }
+
+    return properties;
+  }
+
+  /** serialize computeColumns into properties */
```
Comment (zhoujinsong) [inline flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java:270]:
```suggestion
  /** Serialize compute columns into properties. */
```

File: flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java [hunk 10]
```diff
@@ -197,4 +242,191 @@ public static void addPrimaryKey(
...
+        serialized.put(
+            compoundKey(FLINK_PREFIX, COMPUTED_COLUMNS, computeIndex, EXPR),
+            ((TableColumn.ComputedColumn) column).getExpression());
+        computeIndex++;
+      }
+    }
+    return serialized;
+  }
+
+  /** deserialize computeColumns from properties */
```
Comment (zhoujinsong) [inline flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java:292]:
```suggestion
  /** Deserialize compute columns from properties. */
```

File: flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java [hunk 11]
```diff
@@ -197,4 +242,191 @@ public static void addPrimaryKey(
...
+  private static TableColumn deserializeComputeColumn(
+      Map<String, String> tableProperties, int index, List<String> fieldNames) {
+    String expr = tableProperties.get(compoundKey(FLINK_PREFIX, COMPUTED_COLUMNS, index, EXPR));
+    if (!isExprContainField(expr, fieldNames)) {
+      throw new IllegalStateException(
+          "expression " + expr + " do not match any columns in amoro. ");
```
Comment (zhoujinsong) [inline flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java:298]:
```suggestion
          "expression " + expr + " do not match any columns in the table. ");
```

File: flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java [hunk 12]
```diff
@@ -197,4 +242,191 @@ public static void addPrimaryKey(
...
+            expr);
+    return column;
+  }
+
+  private static boolean isExprContainField(String expr, List<String> fieldNames) {
+    if (expr.equalsIgnoreCase(PROCTIME_FUNCTION)) {
+      return true;
+    }
+    for (String fieldName : fieldNames) {
+      if (expr.contains("`" + fieldName + "`")) {
```
Comment (zhoujinsong) [inline flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java:318]: Is charactor ` always required in expr?

File: flink/flink-common/src/main/java/com/netease/arctic/flink/table/DynamicTableFactory.java [hunk 13]
```diff
@@ -124,8 +126,13 @@ public DynamicTableSource createDynamicTableSource(Context context) {
       objectPath = new ObjectPath(identifier.getDatabaseName(), identifier.getObjectName());
     }
 
+    // update computed columns and watermark to properties
+    Map<String, String> properties = options.toMap();
```
Comment (zhoujinsong) [inline flink/flink-common/src/main/java/com/netease/arctic/flink/table/DynamicTableFactory.java:130]: Based on the implementation of `ArcticCatalog#getTable`, the `CatalogTable#options` already contains the complete properties. What is the reason for generating them again here?

**Review by zhoujinsong — COMMENTED** [review]
I left some comments, please task another look.

File: flink/flink-common/src/main/java/com/netease/arctic/flink/table/DynamicTableFactory.java (hunk 13 above)
Comment (xieyi888) [reply flink/flink-common/src/main/java/com/netease/arctic/flink/table/DynamicTableFactory.java]: > Based on the implementation of `ArcticCatalog#getTable`, the `CatalogTable#options` already contains the complete properties. What is the reason for generating them again here?

Yes, It seems no need to generate properties.

File: flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java [hunk 14]
```diff
@@ -18,49 +18,97 @@
...
+    // add primary key
     if (CollectionUtils.isNotEmpty(primaryKeys)) {
       builder.primaryKey(primaryKeys.toArray(new String[0]));
     }
+
+    HashSet<Integer> computeIndex = getComputeIndex(tableProperties);
+    List<String> fieldNames = rowType.getFieldNames();
+
+    // add computed columns
+    for (int index : computeIndex) {
```
[the remark this answers is not in the document]
Comment (xieyi888) [reply flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java:104]: > There doesn't seem to be a way to guarantee adding the compute fields to the schema in index order here.

Thanks, use TreeSet to keep order

File: flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java (hunk 12 above)
Comment (xieyi888) [reply flink/flink-common/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java:319]: > Is charactor ` always required in expr?

charactor ` will be keeped in property for expresion during deserialize compute column from Flink TableSchema.

considering the situation below
```         
arctic_catalog.test_db.test_table( Id INT);
```
properity
```
flink.schema.column.1.expr=`oldId`+5
```
whitout charactor `,  compute column for index  1 would be deserialized into compute colum  mistakenly.

**Review by zhoujinsong — APPROVED** [review]
LGTM.
Thanks a lot for your contribution!

**Pull request merged**